src/core/public/workspace/workspaces_client.ts

```diff
@@ -38,33 +39,103 @@ type IResponse<T> =
  * @public
  */
 export class WorkspacesClient {
-  private http: HttpStart;
-  constructor(http: HttpStart) {
+  private http: HttpSetup;
+  public currentWorkspaceId$ = new BehaviorSubject<string>('');
+  public workspaceList$ = new Subject<WorkspaceAttribute[]>();
+  constructor(http: HttpSetup) {
     this.http = http;
+    /**
+     * Add logic to check if current workspace id is still valid
+     * If not, remove the current workspace id and notify other subscribers
+     */
+    this.workspaceList$.subscribe(async (workspaceList) => {
+      const currentWorkspaceId = await this.currentWorkspaceId$.getValue();
+      if (currentWorkspaceId) {
+        const findItem = workspaceList.find((item) => item.id === currentWorkspaceId);
+        if (!findItem) {
+          /**
+           * Current workspace is staled
+           */
+          this.currentWorkspaceId$.next('');
+        }
+      }
+    });
+
+    /**
+     * Initialize workspace list
+     */
+    this.updateWorkspaceListAndNotify();
   }
 
+  private catchedFetch = async <T extends IResponse<any>>(
+    path: string,
+    options: HttpFetchOptions
+  ) => {
+    try {
+      return await this.http.fetch<T>(path, options);
+    } catch (error: unknown) {
+      if (error instanceof HttpFetchError || error instanceof Error) {
+        return {
+          success: false,
+          error: error.message,
+        } as T;
+      }
+
+      return {
+        success: false,
+        error: 'Unknown error',
+      } as T;
+    }
+  };
+
   private getPath(path: Array<string | undefined>): string {
     return resolveUrl(`${WORKSPACES_API_BASE_URL}/`, join(...path));
   }
 
+  private async updateWorkspaceListAndNotify(): Promise<void> {
+    const result = await this.list({
+      perPage: 999,
+    });
+
+    if (result.success) {
+      this.workspaceList$.next(result.result.workspaces);
+    }
+  }
+
   public async enterWorkspace(id: string): Promise<IResponse<null>> {
-    return this.http.post(this.getPath(['_enter', id]));
+    const workspaceResp = await this.get(id);
+    if (workspaceResp.success) {
+      this.currentWorkspaceId$.next(id);
+      return {
+        success: true,
+        result: null,
+      };
+    } else {
+      return workspaceResp;
+    }
   }
 
   public async exitWorkspace(): Promise<IResponse<null>> {
-    return this.http.post(this.getPath(['_exit']));
+    this.currentWorkspaceId$.next('');
```

Comment:
Q: do we still need this function`exitWorkspace` if user always has a workspace?

Reply:
"Whether user always has a workspace" is still under discussion, there is no harm that we keep this API for now I think.

Reply:
Makes sense, thanks